packages/application-config/src/types.ts

```diff
@@ -52,3 +52,107 @@ export type LoadingConfigOptions = {
   processEnv: NodeJS.ProcessEnv;
   applicationPath: string;
 };
+
+// Utility types
+type WordSeparators = '_' | '-';
+
+/**
+Verifies if input string is convertible to number
+
+@example
+```
+const num: ConvertibleToNumber<"80"> = "80"
+```
+*/
+type ConvertibleToNumber<T extends string> =
+  /* eslint-disable @typescript-eslint/no-unused-vars */
+  T extends `${infer _Result extends number}` ? T : never;
+
+/**
+Represents an array of strings split using a given character or character set.
+Underscore characters are omitted.
+
+source: https://github.com/sindresorhus/type-fest/blob/fedbc441a314c1f9f5f6225c993860d0886261da/source/split.d.ts#L22:L29
+
+@example
+```
+declare function split<S extends string, D extends string>(string: S, separator: D): Split<S, D>;
+type Item = 'foo' | 'bar' | 'baz' | 'waldo';
+const items = 'foo,bar,baz,waldo';
+let array: Item[];
+array = split(items, ',');
+```
+*/
+export type Split<
+  S extends string,
+  Delimiter extends string
+> = S extends `${infer Head}${Delimiter}${infer Tail}`
+  ? [
+      Delimiter extends '_'
+        ? `${Head}_`
+        : Delimiter extends '-'
+        ? Tail extends ConvertibleToNumber<Tail>
+          ? `${Head}/`
+          : Head
+        : Head,
+      ...Split<Tail, Delimiter>
+    ]
```

Comment:
I gave it another shot and I managed to get it to a state where we are not forced to change the way we format permission names due to TS complexity. As far as I tested it, it seems that the corresponding transformation is done now in JS and in the type.
<img width="480" alt="Screenshot 2022-10-04 at 08 46 07" src="https://user-images.githubusercontent.com/49066275/193758782-af4d247f-44c4-406e-8a48-14ca1e5c9c1f.png">

Reply:
Thanks for looking into it! It does look promising even though we're stepping foot into "unknown/complex" TS territory here.

I would try to keep things as simple as possible as a first step, for instance only allowing to use `-` and for the value to be lowercase.

If we get more specific requirements from customers we can always update the validation as needed.

Reply:
Sorry, it took me a while to wrap up other things.
I'm not sure if I understand - so you wish to use the following function for formatting the permission name part?
```ts
/**
 * The function formats the `permissionName` to a resource access key.
 * It makes the first character of the string and the next character after a special character an uppercase.
 *
 * @example
 * - avengers --> Avengers
 * - the-avengers --> TheAvengers
 * - avengers-01 --> Avengers01 [ℹ️ mind the difference here compared to `formatEntryPointUriPathToResourceAccessKey`]
 */
const formatPermissionNameToResourceAccessKey = (permissionName: string) =>
  permissionName
    // Each word is split by a hyphen.
    .split('-')
    .map(upperFirst)
    .join('');
```
Or not even map with `upperFirst`?

Reply:
I refactored it a little to use this helper for the additional resource access part:
```ts
const formatPermissionNameToResourceAccessKey = (permissionName: string) =>
  permissionName
    // Each word is split by a hyphen.
    .split('-')
    .map(upperFirst)
    .join('');
```
I'm not really sure how to enforce
> for the value to be lowercase

as you suggested. Should each split word be mapped to lowercase before mapping with `upperFirst`?